commit/report.go

```diff
@@ -22,73 +24,57 @@ import (
-	}
-
-	lggr.Infow("generating report",
-		"roots", outcome.MerkleRootOutcome.RootsToReport,
-		"tokenPriceUpdates", outcome.TokenPriceOutcome.TokenPrices,
-		"gasPriceUpdates", outcome.ChainFeeOutcome.GasPrices,
-		"rmnSignatures", outcome.MerkleRootOutcome.RMNReportSignatures,
-	)
-
+// buildOneReport is the common logic for building a report. Different report building
+// algorithms can reassemble reports by selecting which blessed and unblessed merkle
+// roots to include in the report, and which price updates and rmn signatures to use.
+func buildOneReport(
+	ctx context.Context,
+	lggr logger.Logger,
+	reportCodec cciptypes.CommitPluginCodec,
+	transmissionSchedule *ocr3types.TransmissionSchedule,
+	outcomeType merkleroot.OutcomeType,
+	blessedMerkleRoots []cciptypes.MerkleRootChain,
+	unblessedMerkleRoots []cciptypes.MerkleRootChain,
+	rmnSignatures []cciptypes.RMNECDSASignature,
+	rmnRemoteFSign uint64,
+	priceUpdates cciptypes.PriceUpdates,
+) (*ocr3types.ReportPlus[[]byte], error) {
 	var (
 		rep     cciptypes.CommitPluginReport
 		repInfo cciptypes.CommitReportInfo
 	)
 
-	blessedMerkleRoots := make([]cciptypes.MerkleRootChain, 0)
-	unblessedMerkleRoots := make([]cciptypes.MerkleRootChain, 0)
-
-	for _, r := range outcome.MerkleRootOutcome.RootsToReport {
-		if outcome.MerkleRootOutcome.RMNEnabledChains[r.ChainSel] {
-			blessedMerkleRoots = append(blessedMerkleRoots, r)
-		} else {
-			unblessedMerkleRoots = append(unblessedMerkleRoots, r)
-		}
-	}
-
 	// MerkleRoots and RMNSignatures will be empty arrays if there is nothing to report
 	rep = cciptypes.CommitPluginReport{
 		BlessedMerkleRoots:   blessedMerkleRoots,
 		UnblessedMerkleRoots: unblessedMerkleRoots,
-		PriceUpdates: cciptypes.PriceUpdates{
-			TokenPriceUpdates: outcome.TokenPriceOutcome.TokenPrices.ToSortedSlice(),
-			GasPriceUpdates:   outcome.ChainFeeOutcome.GasPrices,
-		},
-		RMNSignatures: outcome.MerkleRootOutcome.RMNReportSignatures,
+		PriceUpdates:         priceUpdates,
+		RMNSignatures:        rmnSignatures,
 	}
 
-	if outcome.MerkleRootOutcome.OutcomeType == merkleroot.ReportEmpty {
+	if outcomeType == merkleroot.ReportEmpty {
 		rep.BlessedMerkleRoots = []cciptypes.MerkleRootChain{}
 		rep.UnblessedMerkleRoots = []cciptypes.MerkleRootChain{}
 		rep.RMNSignatures = []cciptypes.RMNECDSASignature{}
 	}
 
-	if outcome.MerkleRootOutcome.OutcomeType == merkleroot.ReportGenerated {
+	if outcomeType == merkleroot.ReportGenerated {
 		allRoots := append(blessedMerkleRoots, unblessedMerkleRoots...)
 		sort.Slice(allRoots, func(i, j int) bool { return allRoots[i].ChainSel < allRoots[j].ChainSel })
 		repInfo = cciptypes.CommitReportInfo{
-			RemoteF:     outcome.MerkleRootOutcome.RMNRemoteCfg.FSign,
+			RemoteF:     rmnRemoteFSign,
 			MerkleRoots: allRoots,
-			TokenPrices: rep.PriceUpdates.TokenPriceUpdates,
 		}
 	}
+	// in case of a price-only update, add prices regardless of outcome type.
+	repInfo.PriceUpdates = rep.PriceUpdates
```

Comment:
Pulled in from https://github.com/smartcontractkit/chainlink-ccip/pull/621, this type is used by the contract writer for Solana.